### DISCUSSION TITLE
Running Flux on Argonne machines

### DISCUSSION THREAD

**Discussion by jameshcorbett** [flux-framework/flux-core#3538] [Q&A]
I am trying to get a Flux instance up and running on Theta, an ALCF machine that uses Cobalt as its resource manager. After spack-installing ``flux-sched@master``, I tried to start a multi-node Flux instance:

```
corbett8@thetamom2:/gpfs/mira-home/corbett8> aprun -n 2 -N 1 -d 64 -j 4 -cc depth flux start flux mini run -n1 hostname
sh: ldconfig: command not found
sh: ldconfig: command not found
nid00000
nid00001
```

(The `-N` option determines the number of tasks per node, not the number of allocated nodes. And the `-d 64 -j 4 -cc depth` says to give Flux access to all of the hardware threads on the node. The KNL nodes have 256 hardware threads per node.)

It seems that Flux isn't picking up the fact that it's been launched under MPI---it looks like I get two independent Flux instances. It also seems that Flux isn't registering all of the resources available to it:

```
corbett8@thetamom2:/gpfs/mira-home/corbett8> aprun -n 2 -N 1 -d 64 -j 4 -cc depth flux start flux mini run -n2 -c16 hostname
sh: ldconfig: command not found
sh: ldconfig: command not found
0.048s: job.exception type=alloc severity=0 unsatisfiable request
0.047s: job.exception type=alloc severity=0 unsatisfiable request
2021-02-24T00:46:44.975130Z broker.err[0]: rc2.0: flux mini run -n2 -c16 hostname Exited (rc=1) 3.2s
2021-02-24T00:46:44.977393Z broker.err[0]: rc2.0: flux mini run -n2 -c16 hostname Exited (rc=1) 3.2s
Application 22467445 exit codes: 1
Application 22467445 resources: utime ~8s, stime ~18s, Rss ~18532, inblocks ~1232, outblocks ~0
```

Any idea what might be going on here? How could I help you find the source of the issue?

In the meanwhile I should be able to make good progress even with Flux instances that can only see part of a single node's resources.

**Comment by dongahn** [comment]
@jameshcorbett: try

```
flux resource list
```

And see what this version reports.

It could be that `aprun` binds flux brokers to a subset of cores, which keeps flux from discovering all resources.

If this is the case, we need to turn off binding.

**Reply by jameshcorbett** [reply]
It looks like I just misunderstood the `aprun` options I was using. I fixed them and now Flux sees all 64 cores (the KNL nodes have 4-way SMT):

```
corbett8@thetamom2:/gpfs/mira-home/corbett8> aprun -n 1 -N 1 -d 256 -j 4 -cc depth flux start flux resource list
sh: ldconfig: command not found
     STATE NNODES   NCORES    NGPUS NODELIST
      free      1       64        0 nid00000
 allocated      0        0        0 
      down      0        0        0 
Application 22468226 resources: utime ~3s, stime ~9s, Rss ~17100, inblocks ~586, outblocks ~0
```

**Reply by dongahn** [reply]
Cool. But the bootstrapping (see below) is still an issue?

**Reply by dongahn** [reply]
> sh: ldconfig: command not found

Probably red herring. But where does it come from?

**Reply by jameshcorbett** [reply]
Yeah, it is still an issue. And how could I tell where the ldconfig message is coming from?

**Reply by dongahn** [reply]
`which ldconfig`. Is your standard environment be able to locate it at least? 

> And how could I tell where the ldconfig message is coming from?

I think we will have to do some debugging. But I think you probably want to resolve the PMI issues (below) first...

**Reply by jameshcorbett** [reply]
Oh I thought you meant what application (aprun, flux, or something else) was generating that message. There's no ``ldconfig`` in my PATH.

**Comment by dongahn** [comment]
> It seems that Flux isn't picking up the fact that it's been launched under MPI---it looks like I get two independent Flux instances.

We also need to debug this. Can you set `FLUX_PMI_DEBUG=1` to see if this cause flux to print out PMI debug traces?

**Reply by dongahn** [reply]
Related, could you also create a simple MPI hello world and where it resolves `libpmi`? On the execution, you can run `ldd ./hello` Make sure you can launch/bootrap ./hello with aprun though.

**Reply by jameshcorbett** [reply]
```
corbett8@thetamom2:/gpfs/mira-home/corbett8> FLUX_PMI_DEBUG=1 aprun -n 2 -N 1 -d 256 -j 4 -cc depth flux start flux mini run -n1 hostname
sh: ldconfig: command not found
pmi-debug-singleton[-1]: init = operation completed successfully
pmi-debug-singleton[0]: get_params (rank=0 size=1 kvsname=singleton) = operation completed successfully
pmi-debug-singleton[0]: kvs_get (kvsname=singleton key=flux.instance-level value=<none>) = operation failed
pmi-debug-singleton[0]: finalize = operation completed successfully
sh: ldconfig: command not found
pmi-debug-singleton[-1]: init = operation completed successfully
pmi-debug-singleton[0]: get_params (rank=0 size=1 kvsname=singleton) = operation completed successfully
pmi-debug-singleton[0]: kvs_get (kvsname=singleton key=flux.instance-level value=<none>) = operation failed
pmi-debug-singleton[0]: finalize = operation completed successfully
nid00000
nid00001
Application 22468227 resources: utime ~8s, stime ~20s, Rss ~18688, inblocks ~1798, outblocks ~0
```

**Reply by dongahn** [reply]
I don't this flux should have used "pmi-debug-singleton". Probably it wasn't able to find the system provided PMI. Please see https://github.com/flux-framework/flux-core/discussions/3538#discussioncomment-399703

**Reply by jameshcorbett** [reply]
I confirmed the ability to launch ``./hello`` with ``aprun``. First I got

```
corbett8@thetamom2:/gpfs/mira-home/corbett8> ldd ./hello
	not a dynamic executable
```

because it turns out, according to the [documentation](https://www.alcf.anl.gov/support-center/theta/compiling-and-linking-overview-theta-thetagpu),

> The standard mechanism of linking on Theta is to use compiler wrappers cc, CC, or ftn. The wrappers are aware of the required system libraries and include those automatically. All programming environments on Theta default to static linking.

After changing to dynamic linking:

```
corbett8@thetalogin4:~> ldd ./hello | grep pmi
	libpmi.so.0 => /opt/cray/pe/lib64/libpmi.so.0 (0x00007fd101ad1000)
```

**Reply by dongahn** [reply]
OK. Could you set that path in an environment like the following -- in a way that your `flux start` can see and then check if this makes any difference?

```
PMI_LIBRARY=/opt/cray/pe/lib64/libpmi.so 
```

**Reply by jameshcorbett** [reply]
```
corbett8@thetamom1:/gpfs/mira-home/corbett8> PMI_LIBRARY=/opt/cray/pe/lib64/libpmi.so.0 FLUX_PMI_DEBUG=1 aprun -n 2 -N 1 -d 256 -j 4 -cc depth flux start flux mini run -n1 hostname
flux-broker: pmi-debug-dlopen: library name /opt/cray/pe/lib64/libpmi.so.0
flux-broker: pmi-debug-dlopen: library name /opt/cray/pe/lib64/libpmi.so.0
pmi-debug-dlopen[-1]: init = operation completed successfully
pmi-debug-dlopen[0]: get_params (rank=0 size=2 kvsname=22468287) = operation completed successfully
pmi-debug-dlopen[0]: kvs_get (kvsname=22468287 key=flux.instance-level value=<none>) = operation failed
pmi-debug-dlopen[0]: kvs_put (kvsname=22468287 key=0 value=S1@Qqw*:U%sG!9k3vv!my>jU]/EMfGlJ:*@Ph8>l,tcp://10.236.16.120:49152) = operation failed
flux-broker: broker_pmi_kvs_put: operation failed
flux-broker: bootstrap failed
[NID 00020] 2021-02-24 02:19:36 Apid 22468287: initiated application termination
pmi-debug-dlopen[-1]: init = operation completed successfully
pmi-debug-dlopen[1]: get_params (rank=1 size=2 kvsname=22468287) = operation completed successfully
pmi-debug-dlopen[1]: kvs_get (kvsname=22468287 key=flux.instance-level value=<none>) = operation failed
pmi-debug-dlopen[1]: kvs_put (kvsname=22468287 key=1 value=gwh}-PU1WMLzTE#9>!Vt4mbe4(k?:w{dEOUYaE%%) = operation failed
flux-broker: broker_pmi_kvs_put: operation failed
flux-broker: bootstrap failed
Application 22468287 exit codes: 1
Application 22468287 resources: utime ~0s, stime ~4s, Rss ~7340, inblocks ~208, outblocks ~0
```

**Reply by dongahn** [reply]
Maybe due to PMI version mismatch and/or Cray PMI has non standard stuff. This is a bit beyond me. @garlick and @SteVwonder: any thoughts?

**Reply by garlick** [reply]
Interesting.  We are up and communicating with PMI because get_params worked and gave us reasonable values.  The kvs_get is expected to fail when flux is launched by a foreign resource manager.  But the kvs_put ought to have succeeded.  It's using the provided kvsname.  The value is not very long.  Hmmm.

**Reply by SteVwonder** [reply]
> `kvs_put (kvsname=22468287 key=0 value=S1@Qqw*:U%sG!9k3vv!my>jU]/EMfGlJ:*@Ph8>l,tcp://10.236.16.120:49152)`

Looks like a bunch of gibberish followed by a useful tcp endpoint.  Is the beginning bit garbage?  Or is it encrypted? Weird.

**Reply by dongahn** [reply]
@jameshcorbett: what specific Cray/HPE system is this exactly? If this is an old Cray, I wonder if there is a specific vendor extension to PMI such a way that the user software (e.g., Cray MPI) needs to comply with additionally. If that is the case, we should probably wrap that up into what akin to pmi-shim on CORAL: https://flux-framework.readthedocs.io/en/latest/coral.html as our compat layer w/ Flux. 

@SteVwonder: do you still have your note on your exchanges with Cray on this?

**Reply by garlick** [reply]
> Looks like a bunch of gibberish followed by a useful tcp endpoint. Is the beginning bit garbage? Or is it encrypted? Weird

It's a [Z85](https://rfc.zeromq.org/spec/32/) encoded CURVE key followed by the endpoint.  Together they are the broker's "business card" (borrowed that concept from MPICH).   I'll accept weird though.

**Reply by SteVwonder** [reply]
> @SteVwonder: do you still have your note on your exchanges with Cray on this?

Sadly the notes I have boil down to a random link to a gitlab repo with a Cray PMI header in it (not sure if it was ever intended to be released like that, so I'm hesitant to post it here) and a verbal exchange with Larry where he said the person that dealt with XC is no longer at Cray.  I could easily be wrong about this, but my understanding was that the Cray-specific extensions were new interfaces to PMI as opposed to changing the semantics of existing PMI interfaces.  So IIUC, Cray MPI will only bootstrap under a Cray-enabled launcher, but a Cray-enabled launcher can launch any MPI (not just Cray MPI).  Even if all that is true, it could still be the case that the PMI implementation provided on the system is broken in some way.

> Looks like a bunch of gibberish followed by a useful tcp endpoint. Is the beginning bit garbage?

Comments in the broker PMI code state:  "Each broker writes a "business card" consisting of (currently): pubkey[,URI].  The URI and separator are omitted if broker is a leaf in the TBON and won't be creating its own endpoint."  And that pubkey is the output of `zcert_public_txt` which outputs the public key side of the public/private pair.  So that value actually makes sense.  Please ignore my comment about that 🤦

**Comment by jameshcorbett** [comment]
After talking with @dongahn and @SteVwonder yesterday, it turns out installing Flux on Argonne's machines is really not that important---at least for me. The collaborations I was hoping to do on Argonne machines I can do on LC machines instead. So I am just going to forget about these issues for now.

**Comment by dongahn** [comment]
For the record, there are two things that should be resolved properly before being able to support that platform. This way, we can focus on these issues when we circle back to this.

1. ANL's Conda packaging and Flux's Spack packaging can interplay badly if we are not careful enough. We found ways to mix two. At least we need to document this in platform specific sections like https://flux-framework.readthedocs.io/en/latest/coral.html

2. Cray/HPE PMI is a big big issue. It has a couple of vendor-proprietary extensions and I suspect two problems we see arise from this:
- Failure to bootstrap multi-node flux instance with `aprun`. Will need more investigation.
- Failure to `flux mini run` an application compiled/linked with Cray MPI (in our case we try with `mpi4py`. My hope in getting the info from the vendor is kind of low because of our conversation with the vendor folks the other day. My proposal would be similar to what we did on CORAL: 1) enable low performing MPI solution first; 2) then enable high performing one including MVAPICH and finally Cray MPI (if possible).
